src/ode_solver/reduced_order_ode_solver.cpp

```diff
@@ -0,0 +1,290 @@
+        new_residual /= this->dg->right_hand_side.size();
+        this->pcout << " Step length " << step_length << " . Old residual: " << initial_residual << " New residual: " << new_residual << std::endl;
+        for (iline = 0; iline < maxline && new_residual > initial_residual * reduction_tolerance_2; ++iline) {
+            step_length = step_length * step_reduction;
+            this->dg->solution = old_solution;
+            Epetra_Vector epetra_linesearch_reduced_solution_update(epetra_reduced_solution_update);
+            epetra_linesearch_reduced_solution_update.Scale(step_length);
+            epetra_pod_basis.Multiply(false, epetra_linesearch_reduced_solution_update, epetra_solution_update);
+            epetra_solution.Update(1, epetra_solution_update, 1);
+            this->dg->assemble_residual();
+            epetra_test_basis->Multiply(true, epetra_right_hand_side, epetra_reduced_rhs);
+            epetra_reduced_rhs.Norm2(&new_residual);
+            new_residual /= this->dg->right_hand_side.size();
+            this->pcout << " Step length " << step_length << " . Old residual: " << initial_residual << " New residual: " << new_residual << std::endl;
+        }
+    }
+
+    if (iline == maxline) {
+        this->pcout << " Line Search (Case 4): Reverse Search Direction AND Increase nonlinear residual tolerance by a factor " << std::endl;
+        step_length = -1.0;
+        epetra_solution.Update(-1, epetra_solution_update, 1);
+        this->dg->assemble_residual();
+        epetra_test_basis->Multiply(true, epetra_right_hand_side, epetra_reduced_rhs);
+        epetra_reduced_rhs.Norm2(&new_residual);
+        new_residual /= this->dg->right_hand_side.size();
+        this->pcout << " Step length " << step_length << " . Old residual: " << initial_residual << " New residual: " << new_residual << std::endl;
+        for (iline = 0; iline < maxline && new_residual > initial_residual * reduction_tolerance_2; ++iline) {
+            step_length = step_length * step_reduction;
+            this->dg->solution = old_solution;
+            Epetra_Vector epetra_linesearch_reduced_solution_update(epetra_reduced_solution_update);
+            epetra_linesearch_reduced_solution_update.Scale(step_length);
+            epetra_pod_basis.Multiply(false, epetra_linesearch_reduced_solution_update, epetra_solution_update);
+            epetra_solution.Update(1, epetra_solution_update, 1);
+            this->dg->assemble_residual();
+            epetra_test_basis->Multiply(true, epetra_right_hand_side, epetra_reduced_rhs);
+            epetra_reduced_rhs.Norm2(&new_residual);
+            new_residual /= this->dg->right_hand_side.size();
+            this->pcout << " Step length " << step_length << " . Old residual: " << initial_residual << " New residual: " << new_residual << std::endl;
+        }
+    }
+
+    if(iline == maxline){
+        this->pcout << "Line search failed. Returning to old solution." << std::endl;
```

Comment:
I would rather it not abort, as in some very rare cases (usually with very coarse POD bases) the reduced-order solver may fail to converge properly for some points. If I am running an adaptive sampling case, this isn't that big of a deal as it will eventually get computed again and probably converge with a different POD basis. Aborting would cause the adaptive sampling to completely stop, which I don't want.

Reply:
Okay, sounds good!